### ISSUE TITLE
PostgreSQL Performance Improvement Request - Usage of "is not distinct from"

### ISSUE THREAD

**Issue by davidtillfcg** [cap-js/cds-dbs#156]
### Queries to read and update, even by key, use "is not distinct from" which is performance intensive.

> Tested with CAP 7.1.1

> It seems in generating the queries to the DB, this part [here](https://github.com/cap-js/cds-dbs/blob/bfe4749c69f0fb41cfcd10bb0768c611ddd445b2/postgres/lib/PostgresService.js#L295) transforms the '=' operator to a "is not distinct from" statement.

```js
 if (x === '=') return 'is not distinct from'
```

> When running a simple query like

```sh
GET /odata/v4/order/orders?$filter=customerAccountNumber eq '12345' 


Subquery Scan on orders  (cost=22316.11..22316.13 rows=1 width=32) (actual time=174.156..180.259 rows=1 loops=1)
  ->  Limit  (cost=22316.11..22316.11 rows=1 width=94) (actual time=174.133..180.234 rows=1 loops=1)
        ->  Sort  (cost=22316.11..22316.11 rows=1 width=94) (actual time=174.132..180.232 rows=1 loops=1)
              Sort Key: orders_0.order_uuid
              Sort Method: quicksort  Memory: 25kB
              ->  Gather  (cost=1000.00..22316.10 rows=1 width=94) (actual time=174.122..180.224 rows=1 loops=1)
                    Workers Planned: 2
                    Workers Launched: 2
                    ->  Parallel Seq Scan on app_orders_orders orders_0  (cost=0.00..21316.00 rows=1 width=94) (actual time=128.641..163.187 rows=0 loops=3)
                          Filter: (NOT ((customeraccountnumber)::text IS DISTINCT FROM '203998123'::text))
                          Rows Removed by Filter: 346816
Planning Time: 0.185 ms
Execution Time: 180.303 ms
```

For example this table has 1 million records, it's large but not outside of possibilities.  This same query if you simply check field = value is 35ms. 

For those who might say just use keys, or this isn't realistic, note that even using the key, the same query is generated but it has a "limit 1" added to it. So it is still bad, but would depend on where in the table your record is. Same for PATCH/update requests, the update where condition uses is not distinct from. 

### Detailed steps to reproduce

> Setup any CAP project with PostgreSQL with the DB
> Observe the generated SQL query on filtering by a field or by the key

### Suggestion
Would it be possible to consider alternatives that don't cause a table scan? This would help improve performance almost across the board for PostgreSQL.

```sh
WHERE (x = y OR (x IS NULL AND y IS NULL))
```

The above is commonly suggested though it does have some subtle differences, 

```sh
((x IS NOT NULL AND y IS NOT NULL AND x = y) OR (x IS NULL AND y IS NULL))
```

I think the above is probably functionally equivalent but I'm not at all fussed with the how, I just want to avoid a table scan for basic operations. Worst case you could probably identify if you need to care about nulls or not. Similar story for the "!=" I would imagine as well.

Alternatively if there's a reason "is not distinct from" must be used would be good to clarify why, since this code is used across quite a few scenarios and I am only really looking at the very basic.

Please let me know if I can provide any more info or you have any questions.

Also apologies I can't change the labels, it's not technically a bug, more enhancement.

**Comment by BobdenOs** [comment]
@davidtillfcg thank you for pointing out the special behavior `distinct from` invoces in Postgres.

The reason why we use it is that we need to check `=` and `!=` more like javascript then SQL to be OData compliant.

```sql
((x IS NOT NULL AND y IS NOT NULL AND x = y) OR (x IS NULL AND y IS NULL))
```

When using an expression like this it is important to remember that `x` and `y` can be any expression. There for it might mean multiple subselects being printed multiple times. It might be possible to prevent it with something like:
 
```sql
(select ((x IS NOT NULL AND y IS NOT NULL AND x = y) OR (x IS NULL AND y IS NULL)) from (select x,y))
```

My expectation would be that Postgres could optimize queries better when they are expressive and use native functions to achieve a specific behavior. So as `distinct from` is a native feature it should be easier for Postgres to optimize the query. If the filter is done on an not null able column to transform it to `=` or `!=` again.

Postgres is not the only database which has a more expensive `distinct from` logic then just `=`. There for there will at some point be the logic to check for not null able expressions and minimize the usage.

I will see whether it is currently known in some cases that one of the sides of the equation is not null able and if this information is available to switch to `=` and `!=`. Most likely this would then only apply if one of the sides is a table column directly. Like:

```sql
... where ID = 201
```

**Comment by davidtillfcg** [comment]
Thanks for the quick reply, that makes sense. Absolutely other DBMS have adopted "is distinct from" and its behaviour, it's certainly valid to use it.  

I understand it can get quite complex quickly. It'd be great if PostgreSQL would optimise these queries itself but from what I have observed at least with the current version on BTP, it still results in a table scan so is a lot slower than it could be.

Even if the simplest of cases (e.g. filtering a column by a value as above, or on a key) could be identified and handled, this would improve probably the vast majority of use cases especially on relatively large volumes of data.

Looking forward to seeing where this goes, thanks again for the quick response and let me know if you need anything from me.

**Comment by BobdenOs** [comment]
@davidtillfcg I had a look and the information is available to use `=` and `!=` when both sides will never be `NULL` ([Change](https://github.com/cap-js/cds-dbs/compare/cap-js/156?expand=1#diff-0199f4d834e6005bdff85a4c4aaa138aab7c04b8459076e8e1ab263d51d7509eR295-R296)).

I did not measure any difference in performance with our largest example project (SFlight), but the dataset is still only in the thousands not millions of rows. If you are interested you can try out the PR branch and see if it helps for your example.

**Comment by davidtillfcg** [comment]
Hi @BobdenOs , thanks again for the quick action I am glad to see this so well supported.

I tried the PR branch this morning, unfortunately it seemed that for my scenarios the result of operator_has_null always resulted in true and so it still gave the 'is not distinct from' behaviour. 

```js
if (x === '=') return this.operator_has_null(i, xpr) ? 'is not distinct from' : '='
```

I even tried on a key, ie /odata/v4/order/orders(172da28d-3c3a-20e8-0937-22eb45fc7dc9)

Here the field is defined in the entity as key ORDER_UUID: UUID;

In the left/right it gets
```js
['orders','ORDER_UUID']
val: '172da28d-3c3a-20e8-0937-22eb45fc7dc9'
```
Left has key: true, Right has undefined for key and notNull, so
```js
right.key || right.notNull || left != null 
```
is also true.

Not sure if I've missed something. However I did temporarily swap the condition so that it goes down the path of =, the result of a very simple query changes from a fluctuating 300-500ms response to ~60ms pretty constant. With ~30ms overhead in networking, this is really awesome.

I see there's still some discussion going on, but once again thank you for your efforts, keen to see this improvement in place and the possibilities for production usage.
